scalding-commons-macros/src/main/scala/com/twitter/scalding/commons/macros/impl/ordser/UnionLike.scala

```diff
@@ -0,0 +1,279 @@
+ limitations under the License.
+ */
+package com.twitter.scalding.commons.macros.impl.ordser
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.OrderedSerialization
+import com.twitter.scalding.macros.impl.ordser._
+
+object UnionLike {
+
+  def compareBinary(c: Context)(inputStreamA: c.TermName, inputStreamB: c.TermName)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+    val valueA = freshT("valueA")
+    val valueB = freshT("valueB")
+    val idxCmp = freshT("idxCmp")
+
+    val compareSameTypes: Tree = subData.foldLeft(Option.empty[Tree]) {
+      case (existing, (idx, tpe, optiTBuf)) =>
+
+        val commonCmp: Tree = optiTBuf.map{ tBuf =>
+          tBuf.compareBinary(inputStreamA, inputStreamB)
+        }.getOrElse[Tree](q"0")
+
+        existing match {
+          case Some(t) =>
+            Some(q"""
+              if($valueA == $idx) {
+                $commonCmp
+              } else {
+                $t
+              }
+            """)
+          case None =>
+            Some(q"""
+                if($valueA == $idx) {
+                  $commonCmp
+                } else {
+                  sys.error("Unable to compare unknown type")
+                }""")
+        }
+    }.get
+
+    q"""
+        val $valueA: Int = $inputStreamA.readByte.toInt
+        val $valueB: Int = $inputStreamB.readByte.toInt
+        val $idxCmp: Int = $valueA.compare($valueB)
+        if($idxCmp != 0) {
+          $idxCmp
+        } else {
+          $compareSameTypes
+        }
+      """
+  }
+
+  def put(c: Context)(inputStream: c.TermName, element: c.TermName)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val innerArg = freshT("innerArg")
+    subData.foldLeft(Option.empty[Tree]) {
+      case (optiExisting, (idx, tpe, optiTBuf)) =>
+        val commonPut: Tree = optiTBuf.map { tBuf =>
+          q"""val $innerArg: $tpe = $element.asInstanceOf[$tpe]
+              ${tBuf.put(inputStream, innerArg)}
+              """
+        }.getOrElse[Tree](q"()")
+
+        optiExisting match {
+          case Some(s) =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $inputStream.writeByte($idx.toByte)
+              $commonPut
+            } else {
+              $s
+            }
+            """)
+          case None =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $inputStream.writeByte($idx.toByte)
+              $commonPut
+            }
+            """)
+        }
+    }.get
+  }
+
+  def length(c: Context)(element: c.Tree)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): LengthTypes[c.type] = {
```

Comment:
Seems like you want a `SubData` type wraps a head and tail (a non-Empty list). Then I think you could put the foldLeft(None) { }.get pattern in there so it is not repeated.